### PR TITLE
Ad-hoc Git commit versions: support submodules

### PR DESCRIPTION
**Pull request by scheibelp**
Spec versions that reference Git commits (but are not registered as versions in the associated package) had no means to indicate whether submodules should be initialized. This adds logic to fetch submodules if the Package defines a `.submodules` property that returns true for the associated Spec (and also adds a test for this).

TODOs

- [x] This should also probably modify all listed versions in the Package that reference git (for now, if you add a `version` directive  it still has to explicitly set `submodules=True` even if it is specified at the class-definition level).

### CODE REVIEW

File: lib/spack/spack/fetch_strategy.py [hunk 1]
```diff
@@ -1596,7 +1596,13 @@ def for_package_version(pkg, version):
     if version.is_commit and hasattr(pkg, "git"):
         # Populate the version with comparisons to other commits
         version.generate_commit_lookup(pkg.name)
-        fetcher = GitFetchStrategy(git=pkg.git, commit=str(version))
+        kwargs = {
+            'git': pkg.git,
+            'commit': str(version)
+        }
+        if hasattr(pkg, 'submodules') and pkg.submodules:
```
Comment (becker33) [inline lib/spack/spack/fetch_strategy.py]:
```suggestion
        if getattr(pkg, 'submodules'. False):
```

**Review by becker33 — CHANGES_REQUESTED** [review]

**Review by becker33 — APPROVED** [review]

File: var/spack/repos/builtin/packages/axom/package.py [hunk 2]
```diff
@@ -50,6 +50,11 @@ class Axom(CachedCMakePackage, CudaPackage):
     version('0.3.0', tag='v0.3.0', submodules=True)
     version('0.2.9', tag='v0.2.9', submodules=True)
 
+    @property
+    def submodules(self):
```
Comment (white238) [inline var/spack/repos/builtin/packages/axom/package.py:54]: Does this mean all the `submodules=True` above can go away?

Comment (scheibelp) [reply var/spack/repos/builtin/packages/axom/package.py:54]: Yes, I could modify that here if you like

Comment (scheibelp) [reply var/spack/repos/builtin/packages/axom/package.py:54]: I'll make a separate PR for this

**Pull request merged**